extensions/src/libadalang-implementation-extensions.adb

```diff
@@ -650,4 +654,45 @@ package body Libadalang.Implementation.Extensions is
            (Node.Expr_Logic_Vars) (1)'Unrestricted_Access);
    end Expr_P_Type_Var;
 
+   ----------------------------------
+   -- Ada_Node_P_Resolve_Own_Names --
+   ----------------------------------
+
+   function Ada_Node_P_Resolve_Own_Names
+     (Node     : Bare_Ada_Node;
+      Env      : Lexical_Env;
+      Origin   : Bare_Ada_Node;
+      E_Info   : Internal_Entity_Info := No_Entity_Info) return Boolean
+   is
+      R : Relation;
+      C : constant Nameres_Maps.Cursor :=
+        Node.Unit.Nodes_Nameres.Find (Node);
+
+      use Nameres_Maps;
+   begin
+
+      --  There was already resolution for this node, and it's the same
+      --  rebindings, and the cache key is still fresh: just return
+      --  existing result.
+
+      if Nameres_Maps.Has_Element (C)
+        and then Element (C).Cache_Version >= Node.Unit.Context.Cache_Version
+        and then Nameres_Maps.Element (C).Rebindings = E_Info.Rebindings
```

Comment:
Just in case: can it be a problem if we reuse resolution results while the rebindings are the same, but not the node metadata?

Reply:
No, that's why I discriminated on rebindings, not node metadata